3rdParty/iresearch/core/utils/file_utils.cpp

```diff
@@ -768,20 +775,34 @@ handle_t open(const file_path_t path, OpenMode mode, int advice) noexcept {
     }
   } while ((--try_count) > 0);
   return handle_t(nullptr);
-  #else
-    auto fd = ::open(path ? path : "/dev/null", (OpenMode::Read == mode ? O_RDONLY : (O_CREAT | O_TRUNC | O_WRONLY)), S_IRUSR | S_IWUSR);
-    if (fd < 0) {
-      IR_FRMT_ERROR("Failed to open file, error: %d, path: " IR_FILEPATH_SPECIFIER, errno, path);
-      return handle_t(nullptr);
-    }
+#else
+  auto posix_mode =
+      (((OpenMode::Read & mode) == OpenMode::Read) ? O_RDONLY : (O_CREAT | O_TRUNC | O_WRONLY));
+#ifndef __APPLE__
+  if ((mode & OpenMode::Direct) == OpenMode::Direct) {
+    posix_mode |= O_DIRECT;
+  }
+#endif
+  auto fd = ::open(path ? path : "/dev/null", posix_mode, S_IRUSR | S_IWUSR);
+  if (fd < 0) {
+    IR_FRMT_ERROR("Failed to open file, error: %d, path: " IR_FILEPATH_SPECIFIER, errno, path);
+    return handle_t(nullptr);
+  }
+#ifdef __APPLE__
+  if ((mode & OpenMode::Direct) == OpenMode::Direct) {
+    fcntl(fd, F_NOCACHE, 1);
+  }
+#endif
 #if (_XOPEN_SOURCE >= 600 || _POSIX_C_SOURCE >= 200112L) && !defined(__APPLE__)
-    posix_fadvise(fd, 0, 0, advice);
+  posix_fadvise(fd, 0, 0, advice);
 #endif
-    return handle_t(reinterpret_cast<void*>(fd));
+  return handle_t(reinterpret_cast<void*>(fd));
   #endif
 }
 
 handle_t open(void* file, OpenMode mode, int advice) noexcept {
+  assert(mode & (OpenMode::Write | OpenMode::Read) !=
+                    (OpenMode::Write | OpenMode::Read));
```

Comment:
| -- return OpenMode
We don't have such OpenMode
Also what about only Direct?

```suggestion
  assert(((mode & OpenMode::Read) == OpenMode::Invalid) !=
         ((mode & OpenMode::Write) == OpenMode::Invalid));
```